tests/test_nessie_cli.py

```diff
@@ -300,47 +301,55 @@ def test_assign() -> None:
     assert tags["v1.0"] == refs["dev"]
 
 
-@pytest.mark.vcr
+@pytest.mark.nessie
 def test_merge() -> None:
     """Test merge operation."""
+    make_commit("initial_commit", _new_table(), "main", message="Initial commit")
+    initial_hash = ref_hash("main")
     execute_cli_command(["branch", "dev"])
-    make_commit("merge.foo.bar", _new_table("test_merge"), "dev")
+    make_commit("merge_foo_bar", _new_table(), "dev")
     main_hash = ref_hash("main")
     dev_hash = ref_hash("dev")
 
-    # Passing detached commit-id plus a _different_ hash-on-ref --> error
-    execute_cli_command(["merge", f"dev@{dev_hash}", "-c", main_hash, "-o", main_hash], ret_val=1)
     merge_output = execute_cli_command(["merge", "dev", "-c", main_hash])
 
     branches = ReferenceSchema().loads(execute_cli_command(["--json", "branch"]), many=True)
     refs = {i.name: i.hash_ for i in branches}
 
     expected_output_list = [
-        "The following 1 commits were merged onto main:",
-        '{} "test message" '.format(refs["dev"]),
-        "Resultant hash on main after merge: {}".format(refs["main"]),
+        f"Merged dev onto main (was on {main_hash} before merge)",
+        f"Identified merge base commit {initial_hash}",
+        f"Resultant hash on main after merge: {refs['main']}",
     ]
     assert_that(merge_output.splitlines()).is_equal_to(expected_output_list)
 
-    # if we try to merge again from dev to main we get an error
-    # this is because there is now a conflict as the same commit exists on both branches
-    merge_error = execute_cli_command(["merge", "dev", "-c", refs["main"]], ret_val=1)
-    assert "Client Error Conflict: The following keys have been changed in conflict:" in merge_error
+    print(merge_output.splitlines())
+
+    # If we try to merge again from dev to main we get an error.
+    # This is because there is nothing more to merge.
+    merge_output = execute_cli_command(["merge", "dev", "-c", refs["main"]])
+    expected_output_list = [
+        f"Nothing merged from dev onto main (still on {refs['main']})",
+        f"Identified merge base commit {dev_hash}",
+        f"Current, unchanged hash on main after merge: {refs['main']}",
+    ]
+    assert_that(merge_output.splitlines()).is_equal_to(expected_output_list)
 
     logs = simplejson.loads(execute_cli_command(["--json", "log"]))
     # we don't check for equality of hashes here because a merge
     # produces a different commit hash on the target branch
-    assert len(logs) == 1
-    assert_that(logs[0]["message"]).is_equal_to("test message")
+    assert_that(logs).is_length(2)
+    assert_that(logs[0]["message"]).starts_with("Merged dev at ")
     logs = simplejson.loads(execute_cli_command(["--json", "log", "dev"]))
-    assert len(logs) == 1
+    assert_that(logs).is_length(2)
```

Comment:
nit: pytest has smart asserts that provide full context on failure:

https://docs.pytest.org/en/6.2.x/assert.html

so personally dont see necessity for assert_that usage tbh

Reply:
TIL ;)